### PR TITLE
Feature/ngram basic preprocessor

### PR DESCRIPTION
**Pull request by caiyinqiong**
- add Ngram
- delete DSSMPreprocessor, CDSSMPreprocessor and CDSSMPadding

### CODE REVIEW

File: matchzoo/dataloader/callbacks/ngram.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+            onehot[np.arange(num_ngrams), word_ngram] = 1
+            word_to_ngram[idx] = onehot
+        elif mode == 'sum' or mode == 'arrgegate':
+            onehot = np.zeros((num_ngrams, ngram_size))
+            onehot[np.arange(num_ngrams), word_ngram] = 1
+            sum_vector = np.sum(onehot, axis=0)
+            word_to_ngram[idx] = sum_vector
+        else:
+            raise ValueError(f'mode error, it should be one of `index`, '
+                             f'`onehot`, `sum` or `arrgegate`.'
```
Comment (Chriskuei) [inline matchzoo/dataloader/callbacks/ngram.py]: aggregate

File: tutorials/ranking/arcii.ipynb [hunk 2]
```diff
@@ -3,6 +3,16 @@
   {
    "cell_type": "code",
    "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import sys\n",
+    "sys.path.append('../../../MatchZoo-py/')"
```
Comment (Chriskuei) [inline tutorials/ranking/arcii.ipynb]: Delete the first cell

File: tutorials/ranking/arci.ipynb [hunk 3]
```diff
@@ -3,6 +3,16 @@
   {
    "cell_type": "code",
    "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import sys\n",
+    "sys.path.append('../../../MatchZoo-py/')"
```
Comment (Chriskuei) [inline tutorials/ranking/arci.ipynb]: Delete the first cell

File: tutorials/ranking/drmm.ipynb [hunk 4]
```diff
@@ -11,6 +21,8 @@
      "name": "stdout",
      "output_type": "stream",
      "text": [
+      "/users/chenjiangui/caiyinqiong/fix_padding2/MatchZoo-py/tutorials/ranking\n",
+      "/users/chenjiangui/caiyinqiong/fix_padding2/MatchZoo-py\n",
```
Comment (Chriskuei) [inline tutorials/ranking/drmm.ipynb]: It should not contain any specific server information in tutorials.

File: tutorials/ranking/cdssm.ipynb [hunk 5]
```diff
@@ -9,6 +9,8 @@
      "name": "stdout",
      "output_type": "stream",
      "text": [
+      "/users/chenjiangui/caiyinqiong/fix_padding2/MatchZoo-py/tutorials/ranking\n",
+      "/users/chenjiangui/caiyinqiong/fix_padding2/MatchZoo-py\n",
```
Comment (Chriskuei) [inline tutorials/ranking/cdssm.ipynb]: Same issue

File: matchzoo/auto/preparer/preparer.py [hunk 6]
```diff
@@ -172,18 +169,15 @@ def _build_dataset_builder(self, model, embedding_matrix, preprocessor):
...
-            )
-            builder_kwargs['callbacks'].append(hashing_callback)
+        if isinstance(model, mz.models.CDSSM):
+            triletter_callback = mz.dataloader.callbacks.Ngram(
+                preprocessor, mode='sum')
+            builder_kwargs['callbacks'].append(triletter_callback)
+
+        if isinstance(model, mz.models.DSSM):
+            triletter_callback = mz.dataloader.callbacks.Ngram(
+                preprocessor, mode='arrgegate')
```
Comment (faneshion) [inline matchzoo/auto/preparer/preparer.py]: `mode='aggregate'`

File: matchzoo/dataloader/callbacks/ngram.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        batch_size = len(x['text_left'])
+        x['ngram_left'] = [[] for i in range(batch_size)]
+        x['ngram_right'] = [[] for i in range(batch_size)]
+        for idx, row in enumerate(x['text_left']):
+            for term in row:
+                x['ngram_left'][idx].append(self._word_to_ngram[term])
+        for idx, row in enumerate(x['text_right']):
+            for term in row:
+                x['ngram_right'][idx].append(self._word_to_ngram[term])
+        if self._mode == 'arrgegate':
```
Comment (faneshion) [inline matchzoo/dataloader/callbacks/ngram.py]: aggregate

File: matchzoo/dataloader/callbacks/ngram.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    index_term: dict,
+    mode: str = 'index'
+) -> dict:
+    """
+    Generate the word to ngram vector mapping.
+
+    :param ngram_process_unit: The fitted :class:`NgramLetter` object.
+    :param ngram_vocab_unit: The fitted :class:`Vocabulary` object.
+    :param index_term: The index to term mapping dict.
+    :param mode:  It be one of 'index', 'onehot', or 'sum'.
```
Comment (faneshion) [inline matchzoo/dataloader/callbacks/ngram.py]: add 'aggregate' to the mode.

File: matchzoo/dataloader/callbacks/ngram.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+            ngrams = ngram_process_unit.transform([word])
+            word_ngram = ngram_vocab_unit.transform(ngrams)
+        num_ngrams = len(word_ngram)
+        if mode == 'index':
+            word_to_ngram[idx] = word_ngram
+        elif mode == 'onehot':
+            onehot = np.zeros((num_ngrams, ngram_size))
+            onehot[np.arange(num_ngrams), word_ngram] = 1
+            word_to_ngram[idx] = onehot
+        elif mode == 'sum' or mode == 'arrgegate':
```
Comment (faneshion) [inline matchzoo/dataloader/callbacks/ngram.py]: same as above~

File: matchzoo/dataloader/callbacks/ngram.py (hunk 1 above)
Comment (faneshion) [inline matchzoo/dataloader/callbacks/ngram.py]: same as above~

**Pull request merged**